Add fuzzy-date function to case search

MartinRiese opened the pull request:
## Product Description

Add a function to fuzzy search dates.

## Technical Summary

https://dimagi.atlassian.net/browse/USH-4861

## Feature Flag

Case search needs to be enabled

## Safety Assurance

### Safety story

### Automated test coverage

Added new unit test

### QA Plan

Update Potential Duplicate Clients case list in staging and perform search.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

MartinRiese committed:
Add fuzzy-date function to case search

MartinRiese commented:
@esoergel No tests or documentation yet. But I was curious about your thoughts on this? I was think about adding a third parameter to let you specify a range in years for how far the permutations could be away from the original date.

MartinRiese committed:
Add test and documentation for fuzzy-date

esoergel commented on corehq/apps/case_search/xpath_functions/query_functions.py:
```diff
@@ -95,6 +97,50 @@ def fuzzy_match(node, context):
+
+def date_permutations(date_str):
+    [year, month, day] = date_str.split('-')
+    decade_flipped = year[:2:] + year[2::][::-1]
```
```suggestion
    decade_flipped = year[:2] + year[:1:-1]
```
or 
```suggestion
    decade_flipped = year[:2] + year[3] + year[2]
```

MartinRiese replied:
Agreed. Less colons is better :)

esoergel commented on corehq/apps/case_search/xpath_functions/query_functions.py:
```diff
@@ -95,6 +97,50 @@ def fuzzy_match(node, context):
+        f"{decade_flipped}-{reverse_day}-{month}",
+        f"{decade_flipped}-{reverse_month}-{reverse_day}",
+        f"{decade_flipped}-{reverse_day}-{reverse_month}"
+    ]
```
This list is super readable and makes this function's implementation clear - nice work!

esoergel reviewed (commented):
Implementation looks great, and the profiling results so far sound quite promising

MartinRiese commented on corehq/apps/case_search/xpath_functions/query_functions.py:
```diff
@@ -95,6 +97,50 @@ def fuzzy_match(node, context):
+    property_name = _property_name_to_string(node.args[0], node)
+    value = unwrap_value(node.args[1], context)
+
+    return case_property_query(property_name, date_permutations(value))
```
Just to clarify: you mean I have two cases. One with 2024-12-03 and one with 2024-03-12. If I search for 2024-12-03 I should guaranteed that the first one shows up first/ is ranked higher? Are the case search results sorted by the ES score?

esoergel replied:
Yeah I believe that's the case.  Good to confirm with testing though.

MartinRiese committed:
More readable implementation of flipping decade digits in year

Co-authored-by: Ethan Soergel <esoergel@users.noreply.github.com>

MartinRiese committed:
Rename decade_flipped to reverse_decade to be consistent

MartinRiese committed:
Restructure fuzzy-date query to boost original val

Use should query with two clauses: 1) terms for all permutations and
2) term for just the orignal date will give cases with an exact date
match a higher score.

MartinRiese commented on corehq/apps/es/case_search.py:
```diff
@@ -233,7 +233,7 @@ def multiplex_to_adapter(domain):
 
 
-def case_property_query(case_property_name, value, fuzzy=False, multivalue_mode=None):
+def case_property_query(case_property_name, value, fuzzy=False, multivalue_mode=None, boost_first=False):
```
@esoergel I am feeling a bit iffy about adding yet another flag here. I see two alternatives:
1) just create a separate function `case_property_query_boost_first` and call it from `fuzzy_date`
2) use `_base_property_query` directly in `fuzzy_date` probably renaming it without the underscore in the process.
I am curious, if you have a preference?

MartinRiese committed:
Add validation for date just in case

Should be formatted correctly beforehand though

MartinRiese committed:
Merge branch 'master' into riese/fuzzy_date

MartinRiese commented on corehq/apps/case_search/tests/test_filter_dsl.py:
```diff
@@ -206,6 +206,21 @@ def test_fuzzy_match(self):
+                "2042-03-21",
+                "2042-12-30"
+            ], boost_first=True)
+        )
```
@AddisonDunn is there a test the run ES queries against test data already? I am having trouble finding any.

MartinRiese replied:
@AddisonDunn pushed the test

MartinRiese committed:
Add test that runs ES search for fuzzy-date

MartinRiese commented on corehq/apps/es/tests/test_case_search_es.py:
```diff
@@ -435,6 +436,21 @@ def test_fuzzy_case_property_query(self):
         )
 
+    def test_fuzzy_date(self):
+        print("############## test_fuzzy_date")
```
I do not.
